Source plugin for LiquidHaskell

adinapoli opened the pull request:
This PR is the "MVP" (_Minimum Viable Product_) that allows LiquidHaskell to be used as a compiler plugin and integrated as part of the GHC compilation pipeline. It requires the companion `liquid-fixpoint` PR, see https://github.com/ucsd-progsys/liquid-fixpoint/pull/435.

While this PR could be merged as-it-is, I will try documenting below what is outstanding in terms of "bells & whistles". None of what's below (with the notable exception of the CI failure) should stop merging this.

## Ergonomic adjustments

As said previously, we currently have the plugin living at `Language.Haskell.Liquid.GHC.Plugin`, which makes perfect sense in terms of our internal module hierarchy but it's a bit of a mouthful for users. We should probably have just a `LiquidHaskell` module which re-exports the plugin, so that users can use it just by having `-fplugin=LiquidHaskell` in their cabal files. I will try adding this as extra comments to this PR.

## Documentation

I have added some basic documentation but more polishing is needed, but that can come after merging this initial strand of work. In particular, I have tried to revamp the current `README` to make people aware of the new plugin and the migration path, but you might want to further expand on that.

## CI

While CircleCI is happily passing, AppVeyor seems to be struggling, with a failure similar to the one described [here](https://github.com/commercialhaskell/stack/issues/3765) but none of the proposed fixes actually work. I have seen this in the past while working on the plugin and I band-aided that by adding `-fexternal-interpreter` when building the library. Recently, however, it seems like `liquid-ghc-prim` is failing to build:

```
liquid-ghc-prim> configure (lib)
liquid-ghc-prim> Configuring liquid-ghc-prim-0.8.6.0...
liquid-ghc-prim> build (lib)
liquid-ghc-prim> Preprocessing library for liquid-ghc-prim-0.8.6.0..
liquid-ghc-prim> Building library for liquid-ghc-prim-0.8.6.0..
liquid-ghc-prim> [1 of 5] Compiling GHC.Prim
liquid-ghc-prim> 
liquid-ghc-prim> [;1msrc\GHC\Prim.hs:17:1: [;1merror:[;1m[;1m
liquid-ghc-prim>     Could not find module `GHC.Prim'
liquid-ghc-prim>     There are files missing in the `ghc-prim-0.6.1' package,
liquid-ghc-prim>     try running 'ghc-pkg check'.
liquid-ghc-prim>     Use -v (or `:set -v` in ghci) to see a list of the files searched for.
liquid-ghc-prim> [;1m   |
liquid-ghc-prim> [;1m17 | [;1mimport "ghc-prim" GHC.Prim as Exports
liquid-ghc-prim> [;1m   |[;1m ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
liquid-ghc-prim> 
liquid-ghc-prim> Access violation in generated code when reading 0xfffffffffffffff8
liquid-ghc-prim> 
liquid-ghc-prim>  Attempting to reconstruct a stack trace...
liquid-ghc-prim> 
liquid-ghc-prim>    Frame	Code address
```

 It's not really clear to me if this is a stack-specific issue or just a misconfigured build configuration. I will try playing with it, although the feedback cycle is a bit slow as I can't test this locally (I am on Mac OS X and I don't have a Windows VM lying around). 

## Migration plan

Currently I had to leave the "hardcoded prelude" files, living inside the `include` directory, so that the old executable could work just fine. Having said that, we should lay out a migration plan where we phase out the legacy executable and we clean up the the codebase from all these redundant files.

## Liquid-platform packages

Currently, I have done my best to move virtually all the files from `include` into their `liquid-*` mirror package counterparts. Having said that, there are still packages which haven't been fully tested or completed (for example `liquid-bytestring`) so we might want to spend a bit of time there. Last but not least **we should probably move away from the "mega-repo" approach** as we are doing now but rather have the different `liquid-*` projects into separate repository. I think this will help with versioning and revisioning.

adinapoli committed:
Initial support for GHC 8.10.1

adinapoli committed:
Expose some internals and debug.trace

adinapoli committed:
Experimental Cabal support for CircleCI

adinapoli committed:
Initial Plugin work

This PR creates a 'GHC.Plugin' module hierarchy in the codebase which
will set the stage for the plugin work. Some provisional code is
provided that sets up the plugin architecture, without calling into any
meaningful LH functions.

adinapoli committed:
Thread state around using an IORef

This commit uses an IORef to thread (immutable) state around, so that
things discovered in the typechecking phase are available at the core
pass during plugin compilation.

adinapoli committed:
Try to call processModule from the core pass

adinapoli committed:
Reuse configureDynFlags

adinapoli committed:
Try to fetch BareSpecs from the loadInterfaceHook

adinapoli committed:
Make it build with 8.6.5 as well

adinapoli committed:
Serialise/Deserialise BareSpecs

This commit makes possible to serialise and deserialise BareSpecs from/to
interface files.

adinapoli committed:
Introduce the GhcMonadLike typeclass

This commit introduces an experiment GhcMonadLike typeclass which is
virtually isomorphic to the 'GhcMonad' class but it doesn't require an
instance for 'ExceptionMonad', allowing us to call most of the functions
present in 'GHC.Interface' also in 'CoreM'.

adinapoli committed:
Successfully run LH via a plugin on a small example

adinapoli committed:
Remove call to updLiftedSpec

This commit removes the need of calling `updLiftedSpec` from the plugin,
because we don't need access to serialised specs on disk.

adinapoli committed:
Resolve names earlier

adinapoli committed:
Try passing around the un-optimised core

adinapoli committed:
Build on ghc 8.10 RC1

adinapoli committed:
Experiment with the testsuite

adinapoli committed:
Reverse spec comments position

This commit reverses the position of the spec comments we receive from
the parseHook in the Plugin. This was necessary for some tests to pass
as it seems LiquidHaskell relies on the spec annotations to arrive in a
particular order.

adinapoli committed:
Generate optimised Core sooner

adinapoli committed:
Ignore hie.yaml

adinapoli committed:
Better spec discovery

adinapoli committed:
Try to support multiple spec loading

adinapoli committed:
Experiment with a richer SpecEnv

The key for efficiently loading specs from disk lies in making the
SpecEnv type a bit more sophisticated. This commit tries to do that in
an experimental fashion.

adinapoli committed:
Improve caching

adinapoli committed:
Clean redundant imports

adinapoli committed:
Experimental CircleCI build for the source plugin

adinapoli committed:
When -DUSE_PLUGIN, thread options correctly

adinapoli committed:
Remove un-necessary de-optimisation

adinapoli committed:
Introduce the liquid-prelude package

This commit separates out some general-utility Haskell files pre-shipped
as part of the `include` directory as a standalone `liquid-prelude`
package so that they can be compiled using the plugin.

adinapoli committed:
Reuse allImports and qualifiedImports from GHC.Interface

This commit reuses the two functions 'qualifiedImports' and 'allImports'
from the GHC.Interface module, trying to reduce the gap between this
module and the implementation at GHC.Plugin.

adinapoli committed:
Serialise the lifted Spec, not the bare one

This commit fixes a bug where in the plugin we were serialising and
storing in the SpecEnv /not/ the lifted spec constructed as part of the
call to 'makeGhcSpec', but rather the original, bare one.

This was causing name resolution problems in the refinement phase.

adinapoli committed:
Change some test files to be more source-plugin-test friendly

adinapoli committed:
Update CircleCI job

adinapoli committed:
Restore test execution for 8.10 executable (no plugin)

adinapoli committed:
Do not use the plugin on liquid-prelude if ghc < 8.10

adinapoli committed:
Copy LICENSE inside liquid-prelude

adinapoli committed:
Remove trace calls from hsSpecificationP

adinapoli committed:
Prepare for code commentary

adinapoli committed:
Update liquid-specs with new include files

adinapoli committed:
Experiment with liquid-base

adinapoli committed:
Baby steps into compilation of liquid-base

adinapoli committed:
Try splitting liquid-base into liqui-ghc-prim as well

adinapoli committed:
2 modules left to compile in liquid-base

adinapoli committed:
Trying to make liquid-prelude compile

adinapoli committed:
WIP

adinapoli committed:
Add more files to liquid-ghc-prim and liquid-base

adinapoli committed:
Try the fatSpec approach

adinapoli committed:
Introduce the LiquidLib concept

This commit tries a different design to store the spec into interfaces
(as annotations). Instead of bundling everything up into a single, fat
spec, this commit tries to store the spec plus any dependency-spec that
was necessary to refine that module.

Last but not least we now look for cached specs into the interfaces all
the time; for home modules we use the `HomePackageTable`. This allows us
to get rid of the SpecEnv abstraction altogether.

adinapoli committed:
Remove the SpecEnv

adinapoli committed:
More cleanup

adinapoli committed:
Update CI

adinapoli committed:
More principled handling of BareSpec(s)

This commit tries to introduce a GADT called LiquidSpec indexed by the
different flavours of the specs we might have during the lifecycle of
the plugin, so that we can write appropriate merging functions.

adinapoli committed:
Introduce the liquid-vector package

adinapoli committed:
Script to generate correct ghc env for cabal testing

adinapoli committed:
Add liquid-containers, add GHC.TypeLits to liquid-base

adinapoli committed:
Add liquid-parallel, set liquidhaskell_datadir in CI

adinapoli committed:
Start porting of liquid-bytestring

adinapoli committed:
Try to fix test include imports

adinapoli committed:
Always check a file, even without annotations

adinapoli committed:
Improve robustness of scripts/generate_testing_ghc_env

adinapoli committed:
Try to fix ReflectLib2 test

adinapoli committed:
More principled mergeTargetWithClient

This commit amends how signatures are mappended together by taking into
account assumptions that might override previous signatures.

adinapoli committed:
Tentative fix for scoping exports in wrapped modules

adinapoli committed:
Add re-exported vars to GhcSrc

This commit is similar in spirit to the one that added to giTcs the
re-exported `TyCon`s, in the sense that this commit does the same but
for re-exported identifiers (i.e. GHC's 'Var's), by adding those to the
'giDefVars'. This was necessary as 'makeGhcSpec' will "export" a given
signature in the final liftedSpec only if the relevant symbol is
exported and "shows up" in the giDefVars.

This was necessary for "wrapper modules" that simply re-exports
everything, like the ones in 'liquid-base' or similar.

adinapoli committed:
Provisionally restore liquid-prelude files in include

This commit provisionally re-introduces the `liquid-prelude` Haskell
files inside the `include` directory, so that the Stack testsuite can
detect and use them.

adinapoli committed:
Partially revert work on Liquid.Bare

adinapoli committed:
Fix test line

adinapoli committed:
Port GHC.Plugin to the new Specs API

adinapoli committed:
Rebase develop

adinapoli committed:
Allow bare embeds and cmeasures into a LiftedSpec

This commit changes 'makeLiftedSpec0' so that we allow input 'embeds'
and 'cmeasures' to survive the lifting process. Something to bear in mind
in that probably we would have to adjust the format of the correspondent
types, as it might make no sense to talk about a 'LocSymbol' for
something serialised into an interface file.

adinapoli committed:
Remove with-compiler from cabal.project

adinapoli committed:
Allow aliases into LiftedSpec(s)

adinapoli committed:
Add all user-defined measures in a LiftedSpec

LiquidHaskell allows for two "flavours" of measures:

1. Measures without a definition, where only the signature is given;
2. Measures with a definition, typically lifted from a Haskell function.

In order to check some files, we need both flavours. However, a
LiftedSpec was exporting only functions with a definition, so this
commit adds also the other type to the final 'measures' field of a
LiftedSpec.

adinapoli committed:
Minor reshuffle of assumptions in liquid-ghc-prim.

Putting stuff where they belong.

adinapoli committed:
Introduce a StableName

This commit eschew the use of a NameSet to keep the exported names for a
module. This is because unique-based data structures relies on the
'Unique' of the underlying data structure to be, well, unique, but this
is not guaranteed by GHC for things like rebuilds, for example. By using
this new StableName we ensure that everything that a "wrapper" module
exports gets correctly refined in the target LiftedSpec.

adinapoli committed:
Tentatively export assumptions in a LiftedSpec

adinapoli committed:
Tentatively export all imeasures into a LiftedSpec

adinapoli committed:
Grab all TC pieces from availableTyThings

This commit fixes an oversight in the implementation of
availableTyThings. In particular, in case we matched an 'AvailTC', the
second field would give us the 'Name's of the functions this Typeclass
provides, but we were actually ignoring those, causing some LH signatures
to be dropped from the final 'liftedSigs'.

adinapoli committed:
Do not use qualified imports in mirror packages

adinapoli committed:
Generalise desugarModule

adinapoli committed:
Preserve dvariance in LiftedSpec

adinapoli committed:
Add more mirror modules to fix some tests

adinapoli committed:
Fetch FamInsts from ModGuts for the plugin

adinapoli committed:
Preserve rinstance in LiftedSpec

adinapoli committed:
Turn more tests into proper named modules

I have tried also the approach of adding a 'main' entrypoint to some of
these, but this turned out to be a problem.

adinapoli committed:
Remove redundant import

adinapoli committed:
Tentative solution for conditional deps

This commit offers a somewhat simple solution to the fact the old GHC
interface relied on "conditional loading" of more specs, depending on
some options in the configuration files.

adinapoli committed:
Expose -package Cabal in CI runner

adinapoli committed:
Drop ExitFailure 2 to indicate validation errors

adinapoli committed:
Fix test T1657 and Golden

Update CI to fix golden test

adinapoli committed:
Fix bytestring/Unsafe test

adinapoli committed:
Add more modules to liquid-base

adinapoli committed:
Preserve invariants, and re-arrange some spec bindings

adinapoli committed:
Ignore implementation of cStringLen

adinapoli committed:
Add more granular options to Data.Text.Fusion.Internal

This commit adds two LIQUID options to Data.Text.Fusion.Internal,
namely:

--no-termination
--no-totality

This is due to the fact the old executable was checking a single target
file by "slurping" in all the related dependencies, so the LIQUID
options were applied "holistically". In this example, we had the
'Data.Text.Fusion.Common' test which was importing '.Internal', but the
LIQUID options were set on the former, not the latter, and now due to
the fact the source plugin compiles each module indpendently, we don't
have a way to "propagate" liquid pragmas.

To say things differently, now we behave with regards liquid options to
the same way GHC deals with language extensions pragmas; we can specify
some global ones by passing them as plugin options, or individually, on
a case-by-case, for each file.

adinapoli committed:
Enforce upper bound on tasty

adinapoli committed:
Do not call saveResult for the plugin

This commit introduces a new function called 'checkTargetInfo' that does
virtually the same of 'liquidOne' but it doesn't call 'saveResult'. This
is because the latter is used by the old executable to cache the
file on disk to perfom compilation avoidance, something that the plugin
gets for free via the GHC plugin pipeline.

adinapoli committed:
Always check for .liquid existence

This commit introduces a 'copyFilesCreateParentDirIfMissing' that,
despite the long-winded name, ensures that the ".liquid" directory is
there and is created before calling the actual 'copyFile'.

adinapoli committed:
Improve error message for checkSizeFun

adinapoli committed:
Add System.Exit wrapper module

adinapoli committed:
Add test scripts

adinapoli committed:
Remove head.hackage stuff

Virtually all the packages we nedeed from head.hackage seems to have
been uploaded to standard Hackage by now, so we can remove the need for
head.hackage.

adinapoli committed:
Restore the use of Oleg's GHC Docker image

adinapoli committed:
Ignore compilation artifacts in test

adinapoli committed:
Do not force recomp in plugin

kosmikus committed:
Adjust the lower bound for hashable.

kosmikus committed:
Make script shebangs more portable.

adinapoli committed:
Initial work to support the new executable

adinapoli committed:
Better test scripts, adjust CI

adinapoli committed:
Add more modules to liquid-base

adinapoli committed:
Build liquid-base before running stack test

adinapoli committed:
Fix .circleci/config.yml

stack is fairly picky when it comes to the set of flags and options used
for compilation. If we are not consistent, we might end up running
'stack test' with different flags than 'stack build --fast liquid-base'
which would cause the latter to /not/ be present inside the stack's
ghc-pkg-db database by the time we call the testsuite.

It also add a flag to disable new executable.

adinapoli committed:
Interim fix for interpreted code

This (interim) fix ensure that we can still invoke the plugin during an
interactive session (i.e. ghci).

adinapoli committed:
Prune redundant imports

This commit prunes a lot of redundant imports deriving from the fact now
our GHC.API re-exports a lot of GHC modules which were previously
imported elsewhere.

adinapoli committed:
Hide more packages

adinapoli committed:
Ignore QQ-relate tests

adinapoli committed:
Initial work on the liquid-platform

This commit moves all the executable into a new package named
`liquid-platform`, which is meant to be a "battery-included" package end
users will need to install to have a CLI for liquidhaskell, which can
now be primarily used as a source plugin.

It also handles the CLI args correctly, as per @kosmikus _docet_.

adinapoli committed:
More effective partitioning of ghcArgs / liquidArgs in new executable

Due to the fact we need (for some tests) the ability to silence the GHC
output by passing `-v0`, it was necessary a more accurate distinction
between a ghc argument and a liquid argument.

adinapoli committed:
Try adjusting .circleci/config.yaml

adinapoli committed:
Forward -i to GHC, not LH

adinapoli committed:
More redundant import pruning

adinapoli committed:
Merge branch 'develop' into adinapoli/source-plugin

adinapoli committed:
Remove use of --fast from CI

adinapoli committed:
Minor tweaks to the logging verbosity

This commit changes the default verbosity of some logging lines,
effectively "demoting" some from "normal" to "loud", so that in the
former configuration LH would only print warnings and errors.

On top of this, we also adjust the new binary to print the banner only
in certain circumstances, to conform to the old output as much as
possible.

adinapoli committed:
Trivially safe indication

This commit changes the LIQUID output format to also output the number
of generated constraints.

adinapoli committed:
Remove isInteractive hack

This commit removes the 'isInteractive' function in favour of simpler
logic, where we call 'parseModule' with a "cleaned" DynFlags (without
the plugins) to avoid any recursive call.

adinapoli committed:
Fix FunClashLibLibClient test

This commit replaces `isPrefixOfSym` with a new `isParentModuleOf` in
the 'matchMod' function, making sure we correctly handle cases where
modules have a very similar name but are clearly distinct.

adinapoli committed:
Do not use -fforce-recomp in liquid

adinapoli committed:
Pass -w when running text and bytestring benchmarks

This commit passes the '-w' flag to suppress all the warnings we would
otherwise get when compiling the 'text' and 'bytestring' benchmarks,
which would make the test failure dump a lot of output, making debugging
harder.

adinapoli committed:
Documentation.

adinapoli committed:
Do not duplicate lifted invariants

adinapoli committed:
Tentative fix for Text.Encoding tests

This commit imports the `Data.ByteString.Lazy.Internal` module inside
`Data.Text.Encoding.Fusion` and then adds the `--prune-unsorted` and
`--compilespec` Liquid pragmas to fix the last remaining tests.

It also re-enables `.Fusion` tests so that they would still pass
for the old executable.

adinapoli committed:
Try to fix Windows build.

adinapoli committed:
Remove GHC.Prim to fix Windows builds

adinapoli committed:
Bump version to 0.8.10.1

gridaphobe commented:
Thanks @adinapoli!! I've been arguing for ages that LiquidHaskell should be a GHC plugin but don't have the time anymore to do the work.

adinapoli committed:
AppVeyor - revert multiline command

adinapoli committed:
Add logging verbosity as a Config option

This commit adds a new field to the LH's `Config` data structure,
unsurprisingly-named `loggingVerbosity`, which allows LH to have a
better control over the granularity of logging.

This allows the source plugin to run in `Quiet` mode (which is now the
default) but the executable to be customised to run by default using the
`--normal` mode, to show things like the progress bar.

adinapoli committed:
Switch AppVeyor build to use legacy exe

adinapoli committed:
Export the plugin in LiquidHaskell

This commit re-exports the source plugin under the 'LiquidHaskell'
module, so that users can simply use `-fplugin=LiquidHaskell`.

adinapoli committed:
Use CPP to selectively export the plugin
